<a href="https://colab.research.google.com/github/vvthakral/data-science-bootcamp/blob/main/amazon_recommendation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem Statement

You are given subset of Amazon Transaction data for year 2019. Recommend a second product to the user based on users input (1st product in cart). The recommendation does not need to involve Machine Learning but simply be based on the count of most commonly purchased products (together) from the purchase history in the dataset.To limit the scope, you can target only San Francisco instead of entire USA.


Learnings:<br>
How to handle huge dataset<br>
How to deal with Textual data<br>
How to build prototype recommender system <br>
How to make recommendations more efficient<br>

In [ ]:
import pandas as pd
import numpy as np

# Reading Data

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving amazon_transaction.csv to amazon_transaction.csv


In [ ]:
df = pd.read_csv("amazon_transaction.csv")
df.head(10)

Order ID  ...                        Purchase Address
0   176558  ...            917 1st St, Dallas, TX 75001
1      NaN  ...                                     NaN
2   176559  ...       682 Chestnut St, Boston, MA 02215
3   176560  ...    669 Spruce St, Los Angeles, CA 90001
4   176560  ...    669 Spruce St, Los Angeles, CA 90001
5   176561  ...       333 8th St, Los Angeles, CA 90001
6   176562  ...  381 Wilson St, San Francisco, CA 94016
7   176563  ...        668 Center St, Seattle, WA 98101
8   176564  ...         790 Ridge St, Atlanta, GA 30301
9   176565  ...  915 Willow St, San Francisco, CA 94016

[10 rows x 6 columns]

#Data Understanding

In [ ]:
#Check for nulls and remove the rows
df.isnull().sum()

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

In [ ]:
df.dropna(inplace=True)

In [ ]:
#Get basic stats
df.describe()

Order ID               Product  ...  Order Date  Purchase Address
count     186305                186305  ...      186305            186305
unique    178438                    20  ...      142396            140788
top     Order ID  USB-C Charging Cable  ...  Order Date  Purchase Address
freq         355                 21903  ...         355               355

[4 rows x 6 columns]

In [ ]:
df = df[df['Purchase Address'] != 'Purchase Address']

In [ ]:
df['Order ID'].count() - len(df['Order ID'].unique())

7513

We have 7513 duplicate order id i.e purchase of more than 1 product was made on these order id. 

This is the data we will be using for making recommendations.

In [ ]:
df = df[df['Order ID'].duplicated(keep=False)]
df

Order ID  ...                       Purchase Address
3        176560  ...   669 Spruce St, Los Angeles, CA 90001
4        176560  ...   669 Spruce St, Los Angeles, CA 90001
18       176574  ...      20 Hill St, Los Angeles, CA 90001
19       176574  ...      20 Hill St, Los Angeles, CA 90001
30       176585  ...      823 Highland St, Boston, MA 02215
...         ...  ...                                    ...
186792   259303  ...          106 7th St, Atlanta, GA 30301
186803   259314  ...     241 Highland St, Atlanta, GA 30301
186804   259314  ...     241 Highland St, Atlanta, GA 30301
186841   259350  ...  519 Maple St, San Francisco, CA 94016
186842   259350  ...  519 Maple St, San Francisco, CA 94016

[14649 rows x 6 columns]

#Feature Engineering

In [ ]:
#extract city from address
def get_city(address):
  return address.split(",")[1].strip(" ")

df['city'] = df['Purchase Address'].apply(get_city)
df.head(10)

Order ID  ...           city
3     176560  ...    Los Angeles
4     176560  ...    Los Angeles
18    176574  ...    Los Angeles
19    176574  ...    Los Angeles
30    176585  ...         Boston
31    176585  ...         Boston
32    176586  ...  San Francisco
33    176586  ...  San Francisco
119   176672  ...  New York City
120   176672  ...  New York City

[10 rows x 7 columns]

In [ ]:
#select data for San Francisco only
df = df[df['city']=='San Francisco']
print(f"Number of orders with more than 1 product purchase in San Francisco are {df['city'].count()}")

Number of orders with more than 1 product purchase in San Francisco are 3562


In [ ]:
#Use the apply function to merge products in same order id.
#refer session notebook or the below link
#https://stackoverflow.com/questions/27298178/concatenate-strings-from-several-rows-using-pandas-groupby

In [ ]:
#Task 1
#Merge products with same order id
'''
After merger, products in same order id should be merged into 1 string
eg: Google Phone, Wired Headphones
'''

df = df.groupby(['Order ID'])['Product'].apply(lambda x: ','.join(x)).reset_index()
df.head()

Order ID                                       Product
0   141450       Google Phone,Bose SoundSport Headphones
1   141782   27in FHD Monitor,Bose SoundSport Headphones
2   141795                       iPhone,Wired Headphones
3   141843  AA Batteries (4-pack),AAA Batteries (4-pack)
4   141946         USB-C Charging Cable,Wired Headphones

#Recommendation

In [ ]:
#Task 2
#Make recommendation
'''
Make simple recommendation, just based on count of combinations from previous purchase history.
eg: 
google phone, wired headphones 5 times
google phone, usb type c 10 times
so recommend usb type c cable if customer has google phone in their cart
'''
combinations = {}
for _ in df['Product']:
  p = _.split(',')
  p.sort()
  for i in range(len(p)):
    for j in range(i+1,len(p)):
      if (p[i],p[j]) in combinations:
        combinations[(p[i],p[j])] +=1
      else:
        combinations[(p[i],p[j])]=1

print(combinations)

{('Bose SoundSport Headphones', 'Google Phone'): 52, ('27in FHD Monitor', 'Bose SoundSport Headphones'): 11, ('Wired Headphones', 'iPhone'): 119, ('AA Batteries (4-pack)', 'AAA Batteries (4-pack)'): 17, ('USB-C Charging Cable', 'Wired Headphones'): 60, ('Bose SoundSport Headphones', 'Lightning Charging Cable'): 22, ('Google Phone', 'iPhone'): 2, ('20in Monitor', 'Wired Headphones'): 3, ('Apple Airpods Headphones', 'USB-C Charging Cable'): 13, ('Google Phone', 'USB-C Charging Cable'): 245, ('Google Phone', 'Wired Headphones'): 93, ('USB-C Charging Cable', 'Vareebadd Phone'): 91, ('Lightning Charging Cable', 'USB-C Charging Cable'): 18, ('AA Batteries (4-pack)', 'Macbook Pro Laptop'): 5, ('Lightning Charging Cable', 'iPhone'): 254, ('Apple Airpods Headphones', 'Lightning Charging Cable'): 26, ('Apple Airpods Headphones', 'iPhone'): 96, ('Apple Airpods Headphones', 'Bose SoundSport Headphones'): 14, ('Lightning Charging Cable', 'Wired Headphones'): 39, ('34in Ultrawide Monitor', 'Lightnin

In [ ]:
#Make list out of dictionary
comb_list = list(combinations.items())
print(comb_list)

[(('Bose SoundSport Headphones', 'Google Phone'), 52), (('27in FHD Monitor', 'Bose SoundSport Headphones'), 11), (('Wired Headphones', 'iPhone'), 119), (('AA Batteries (4-pack)', 'AAA Batteries (4-pack)'), 17), (('USB-C Charging Cable', 'Wired Headphones'), 60), (('Bose SoundSport Headphones', 'Lightning Charging Cable'), 22), (('Google Phone', 'iPhone'), 2), (('20in Monitor', 'Wired Headphones'), 3), (('Apple Airpods Headphones', 'USB-C Charging Cable'), 13), (('Google Phone', 'USB-C Charging Cable'), 245), (('Google Phone', 'Wired Headphones'), 93), (('USB-C Charging Cable', 'Vareebadd Phone'), 91), (('Lightning Charging Cable', 'USB-C Charging Cable'), 18), (('AA Batteries (4-pack)', 'Macbook Pro Laptop'), 5), (('Lightning Charging Cable', 'iPhone'), 254), (('Apple Airpods Headphones', 'Lightning Charging Cable'), 26), (('Apple Airpods Headphones', 'iPhone'), 96), (('Apple Airpods Headphones', 'Bose SoundSport Headphones'), 14), (('Lightning Charging Cable', 'Wired Headphones'), 39)

In [ ]:
#sort list based on the number of occurences/combination count
comb_list = sorted(comb_list,key=lambda x:x[1],reverse=True)
print(comb_list)

[(('Lightning Charging Cable', 'iPhone'), 254), (('Google Phone', 'USB-C Charging Cable'), 245), (('Wired Headphones', 'iPhone'), 119), (('Apple Airpods Headphones', 'iPhone'), 96), (('Google Phone', 'Wired Headphones'), 93), (('USB-C Charging Cable', 'Vareebadd Phone'), 91), (('USB-C Charging Cable', 'Wired Headphones'), 60), (('Bose SoundSport Headphones', 'Google Phone'), 52), (('Vareebadd Phone', 'Wired Headphones'), 44), (('Lightning Charging Cable', 'Wired Headphones'), 39), (('Apple Airpods Headphones', 'Lightning Charging Cable'), 26), (('AA Batteries (4-pack)', 'Wired Headphones'), 26), (('AA Batteries (4-pack)', 'Lightning Charging Cable'), 26), (('Apple Airpods Headphones', 'Wired Headphones'), 25), (('AA Batteries (4-pack)', 'USB-C Charging Cable'), 25), (('Bose SoundSport Headphones', 'Lightning Charging Cable'), 22), (('Bose SoundSport Headphones', 'USB-C Charging Cable'), 22), (('AAA Batteries (4-pack)', 'USB-C Charging Cable'), 21), (('Bose SoundSport Headphones', 'Vare

In [ ]:
def recommender(product):
  '''
  code for recommendation
  Simple method: return any product that was purchased with the product in cart
  better approach: return product with highest match count
  '''
  global comb_list
  product = product.lower().strip() # clean,format the user input
  for i in comb_list:
    if i[0][0].lower() == product:
      return i[0][1]
    elif i[0][1].lower() == product:
      return i[0][0]
  return 'We are still upgrading our recommendation system!'

product = 'google phone'
print(recommender(product))

USB-C Charging Cable
